### PR TITLE
Laboratory work #7, Chekmareva Kristina - 21FPL1

### PR DESCRIPTION
**Pull request by kschekmareva**

### CODE REVIEW

**Comment by demid5111** [comment]
nothing to review to this moment

File: lab_7_llm/settings.json [hunk 1]
```diff
@@ -1,8 +1,8 @@
 {
   "parameters": {
-    "model": "",
-    "dataset": "",
-    "metrics": []
+    "model": "dmitry-vorobiev/rubert_ria_headlines",
+    "dataset": "IlyaGusev/gazeta",
+    "metrics": ["bleu", "rouge"]
```
Comment (demid5111) [inline lab_7_llm/settings.json:5]: layout

**Review by demid5111 — CHANGES_REQUESTED** [review]
ok, good progress

**Comment by demid5111** [comment]
nothing to review to this moment

File: lab_7_llm/main.py [hunk 2]
```diff
@@ -74,6 +122,7 @@ def __getitem__(self, index: int) -> tuple[str, ...]:
         Returns:
             tuple[str, ...]: The item to be received
         """
+        return tuple(self._data.iloc[index])
```
Comment (demid5111) [inline lab_7_llm/main.py:125]: why do you need targets in the tuple?

File: lab_7_llm/main.py [hunk 3]
```diff
@@ -103,6 +153,9 @@ def __init__(
             batch_size (int): The size of the batch inside DataLoader
             device (str): The device for inference
         """
+        super().__init__(model_name, dataset, max_length, batch_size, device)
+        self._model = AutoModelForSeq2SeqLM.from_pretrained(model_name)
+        self._tokenizer = AutoTokenizer.from_pretrained(model_name, model_max_length=max_length)
```
Comment (demid5111) [inline lab_7_llm/main.py:158]: you ignore device

File: lab_7_llm/main.py [hunk 4]
```diff
@@ -103,6 +153,9 @@ def __init__(
             batch_size (int): The size of the batch inside DataLoader
             device (str): The device for inference
         """
+        super().__init__(model_name, dataset, max_length, batch_size, device)
+        self._model = AutoModelForSeq2SeqLM.from_pretrained(model_name)
```
Comment (demid5111) [inline lab_7_llm/main.py:157]: where do you switch a model to eval mode

File: lab_7_llm/main.py [hunk 5]
```diff
@@ -111,6 +164,21 @@ def analyze_model(self) -> dict:
         Returns:
             dict: Properties of a model
         """
+        input_tensor = torch.ones((1, self._model.config.encoder.max_position_embeddings), dtype=torch.long)
+        inputs = {"input_ids": input_tensor, "attention_mask": input_tensor}
+        summary_model = summary(self._model, input_data=inputs, decoder_input_ids=input_tensor, verbose=0)
+        print(summary_model)
```
Comment (demid5111) [inline lab_7_llm/main.py:170]: remove prints

File: lab_7_llm/main.py [hunk 6]
```diff
@@ -111,6 +164,21 @@ def analyze_model(self) -> dict:
...
+        info = {
```
Comment (demid5111) [inline lab_7_llm/main.py:171]: immediately return

File: lab_7_llm/main.py [hunk 7]
```diff
@@ -132,6 +201,15 @@ def infer_dataset(self) -> pd.DataFrame:
         Returns:
             pd.DataFrame: Data with predictions
         """
+        loader = DataLoader(batch_size=self._batch_size, dataset=self._dataset)
+
+        all_predictions = [
+            prediction for batch in loader for prediction in self._infer_batch(batch)
+        ]
+
+        results_df = pd.DataFrame(self._dataset.data)
```
Comment (demid5111) [inline lab_7_llm/main.py:210]: 🏅

File: lab_7_llm/main.py [hunk 8]
```diff
@@ -144,6 +222,12 @@ def _infer_batch(self, sample_batch: Sequence[tuple[str, ...]]) -> list[str]:
         Returns:
             list[str]: Model predictions as strings
         """
+        inputs = self._tokenizer.prepare_seq2seq_batch(list(sample_batch[0]), return_tensors="pt", padding=True, truncation=True,
+                                                       max_length=self._max_length)
+        attention_mask = inputs["attention_mask"].to(self._device)
```
Comment (demid5111) [inline lab_7_llm/main.py:227]: why do you pass to device only attention mask?

File: lab_7_llm/main.py [hunk 9]
```diff
@@ -168,3 +252,4 @@ def run(self) -> dict | None:
         Returns:
             dict | None: A dictionary containing information about the calculated metric
         """
+
```
Comment (demid5111) [inline lab_7_llm/main.py:255]: ?

File: lab_7_llm/start.py [hunk 10]
```diff
@@ -2,15 +2,40 @@
...
+    importer.obtain()
+    preprocessor = RawDataPreprocessor(importer.raw_data)
+    analysis = preprocessor.analyze()
+    print(analysis)
+    preprocessor.transform()
+    dataset = TaskDataset(preprocessor.data.head(100))
+    print(len(dataset))
+    print(dataset[1])
+    pipeline = LLMPipeline(settings['parameters']['model'], dataset, max_length=120, batch_size=1, device='cpu')
+    print(pipeline.analyze_model())
```
Comment (demid5111) [inline lab_7_llm/start.py:35]: demonstrate for 6 - infer_sample

**Review by demid5111 — CHANGES_REQUESTED** [review]
you did not address previous comments. for 8 current code is not enough. but there is good progress, please continue

File: lab_7_llm/main.py (hunk 3 above)
Comment (kschekmareva) [reply lab_7_llm/main.py:159]: add device

File: lab_7_llm/service.py [hunk 11]
```diff
@@ -3,6 +3,8 @@
 """
 # pylint: disable=too-few-public-methods, undefined-variable, unused-import, assignment-from-no-return, duplicate-code
 from pathlib import Path
+from fastapi import FastAPI
+from lab_7_llm.main import LLMPipeline
```
[the remark this answers is not in the document]
Comment (kschekmareva) [reply lab_7_llm/service.py]: pycharm suggested where to import from and without these lines the tests are not passed

File: lab_7_llm/main.py (hunk 4 above)
Comment (kschekmareva) [reply lab_7_llm/main.py]: done

File: lab_7_llm/main.py (hunk 5 above)
Comment (kschekmareva) [reply lab_7_llm/main.py]: done

File: lab_7_llm/main.py (hunk 6 above)
Comment (kschekmareva) [reply lab_7_llm/main.py]: done

File: lab_7_llm/main.py (hunk 2 above)
Comment (demid5111) [reply lab_7_llm/main.py:130]: the question is about content, not type

File: lab_7_llm/main.py [hunk 12]
```diff
@@ -168,3 +278,18 @@ def run(self) -> dict | None:
...
         """
+        results_df = pd.read_csv(self.data_path)
+        texts = results_df[ColumnNames.PREDICTION.value]
+        targets = results_df[ColumnNames.TARGET.value]
+        evaluation = {}
+
+        string_metrics = [format(metric) for metric in self._metrics]
+
+        for metr in string_metrics:
+            metric = load(metr, seed=77).compute(predictions=texts, references=targets)
```
Comment (demid5111) [inline lab_7_llm/main.py:289]: load metrics on init phase

File: lab_7_llm/main.py [hunk 13]
```diff
@@ -168,3 +278,18 @@ def run(self) -> dict | None:
...
+            if metr == Metrics.ROUGE.value:
+                evaluation[metr] = metric[Metrics.ROUGE.value + "L"]
```
Comment (demid5111) [inline lab_7_llm/main.py:291]: `Metrics.ROUGE.value + "L"` too much - use string constant with full name

File: lab_7_llm/main.py [hunk 14]
```diff
@@ -144,6 +231,27 @@ def _infer_batch(self, sample_batch: Sequence[tuple[str, ...]]) -> list[str]:
...
+            return_tensors="pt",
+            padding=True,
+            truncation=True,
+            max_length=self._max_length,
+        ).to(self._device)
+
+        if not isinstance(self._model, torch.nn.Module):
+            raise TypeError("Expected self._model to be an instance of torch.nn.Module.")
+
+        output_ids = self._model.generate(
```
Comment (demid5111) [inline lab_7_llm/main.py:245]: forgot to disable gradients

File: requirements.txt [hunk 15]
```diff
@@ -0,0 +1,10 @@
+datasets==3.2.0
+evaluate==0.4.3
+fastapi==0.115.6
+pandas==2.2.3
+pydantic==2.10.5
+rouge-score==0.1.2
+torch==2.5.1
+torchinfo==1.8.0
+tqdm==4.67.1
+transformers==4.48.0
```
Comment (demid5111) [inline requirements.txt:10]: newline

**Review by demid5111 — CHANGES_REQUESTED** [review]
not so much changed, try to finalize and also mind the comments

File: requirements.txt [hunk 16]
```diff
@@ -0,0 +1,7 @@
+datasets==3.2.0
+fastapi==0.115.6
+pandas==2.2.3
+pydantic==2.10.5
+pytest==8.3.4
+torch==2.5.1
+tqdm==4.67.1
```
[the remark this answers is not in the document]
Comment (kschekmareva) [reply requirements.txt]: done

File: lab_7_llm/main.py (hunk 2 above)
Comment (kschekmareva) [reply lab_7_llm/main.py]: yeah, i got it, we really do not need targets, corrected

File: lab_7_llm/main.py (hunk 13 above)
Comment (kschekmareva) [reply lab_7_llm/main.py]: corrected

File: requirements.txt [hunk 17]
```diff
@@ -0,0 +1,11 @@
+datasets==3.2.0
+evaluate==0.4.3
+fastapi==0.115.6
+pandas==2.2.3
+pydantic==2.10.5
+rouge-score==0.1.2
+torch==2.5.1
```
Comment (demid5111) [inline requirements.txt:7]: use torch version that is required in docs

File: requirements.txt [hunk 18]
```diff
@@ -0,0 +1,11 @@
+datasets==3.2.0
+evaluate==0.4.3
+fastapi==0.115.6
+pandas==2.2.3
+pydantic==2.10.5
```
Comment (demid5111) [inline requirements.txt:5]: why do you need it?

File: lab_7_llm/main.py [hunk 19]
```diff
@@ -159,12 +269,29 @@ def __init__(self, data_path: Path, metrics: Iterable[Metrics]) -> None:
...
+        outputs_df = pd.read_csv(self.data_path)
+        summaries = outputs_df[ColumnNames.PREDICTION.value]
+        targets = outputs_df[ColumnNames.TARGET.value]
+        evaluation = {}
+
+        for metr in self._metrics:
+            metric_result = metr.compute(predictions=summaries, references=targets)
+            metric_name = metr.name
+
+            if metric_name == "rouge":
```
Comment (demid5111) [inline lab_7_llm/main.py:292]: use enum value for metrics from core_utils

File: lab_7_llm/main.py [hunk 20]
```diff
@@ -159,12 +269,29 @@ def __init__(self, data_path: Path, metrics: Iterable[Metrics]) -> None:
...
 
     @report_time
     def run(self) -> dict | None:
         """
-        Evaluate the predictions against the references using the specified metric.
+        Evaluate the predictions against the references using the specified metrics.
 
         Returns:
-            dict | None: A dictionary containing information about the calculated metric
+            dict | None: A dictionary containing the computed metric values.
```
Comment (demid5111) [inline lab_7_llm/main.py:281]: discard docstrings

File: lab_7_llm/main.py [hunk 21]
```diff
@@ -74,6 +128,7 @@ def __getitem__(self, index: int) -> tuple[str, ...]:
         Returns:
             tuple[str, ...]: The item to be received
         """
+        return tuple([str(self._data.loc[index, ColumnNames.SOURCE.value])])
```
Comment (demid5111) [inline lab_7_llm/main.py:131]: when creating a tuple from a single element, you just need to remember about a comma, no need in making a list then casted to tuple

File: lab_7_llm/main.py [hunk 22]
```diff
@@ -159,12 +269,29 @@ def __init__(self, data_path: Path, metrics: Iterable[Metrics]) -> None:
             data_path (pathlib.Path): Path to predictions
             metrics (Iterable[Metrics]): List of metrics to check
         """
+        self.data_path = data_path
+        self._metrics = [load(str(item)) for item in metrics]
```
Comment (demid5111) [inline lab_7_llm/main.py:273]: why did you remove seed?

**Review by demid5111 — CHANGES_REQUESTED** [review]
ok. prepare to the remaining comments and make sure CI is green.

File: lab_7_llm/main.py (hunk 22 above)
Comment (kschekmareva) [reply lab_7_llm/main.py]: corrected, i added seed

File: lab_7_llm/main.py (hunk 21 above)
Comment (kschekmareva) [reply lab_7_llm/main.py]: done

File: lab_7_llm/main.py (hunk 20 above)
Comment (kschekmareva) [reply lab_7_llm/main.py]: done

File: lab_7_llm/main.py (hunk 19 above)
Comment (kschekmareva) [reply lab_7_llm/main.py]: done

File: requirements.txt (hunk 17 above)
Comment (kschekmareva) [reply requirements.txt]: done

File: requirements.txt (hunk 18 above)
Comment (kschekmareva) [reply requirements.txt:5]: without this dependency my code run failed, i use it in service.py
